Feature/thrust guide

gaffarelj opened the pull request:
This is the first version of the thrust guide, which can probably be already useful to P&O students as-is.

The more complex part of the guide on how to use rotational dynamics to define the thrust orientation, and how to use the thrust orientation from aerodynamic guidance, is still to be added later on.

DominicDirkx commented:
Nice work on the guide Jeremie! I've made a bunch of small modifications, but overall I think we can push this soon. A few points:

* A Python user is typically not exposed to the settings _classes_, and I replaced this with links/references to the settings factory functions. This is how we're also approached this in other parts of the guide
* I like the idea of putting notes _inside_ the code blocks, this makes it much more flexible for future C++ extensions
* I'm not sure the combined aero/thrust part at the end is right (which say something about the design....). If you use `thrust_direction_from_state_guidance`, the body inertial orientation gets defined, as the thrust direction in a body-fixed frame is defined by the thrust magnitude settings. If you subsequently define the angle of attack through aerodynamic guidance, I think this will override the original orientation. I propose that we leave this part out for now until we can discuss this in more detail. The rest of the guide can be merged as far as I'm concerned. Well done!

gaffarelj commented:
Thank you for your review, and for your improvements.

Concerning the aero/thrust part and example: I have tested this scenario with my (in progress) rocket ascent code. From what I could see, the aerodynamic angles and thrust orientation are kept separate.
Changing the angle of attack from a sine wave to constant values of 0 and -45deg induced changes in the aerodynamic acceleration, but kept the trajectory more or less the same. 
I conclude from this that the thrust orientation was not impacted and that both of the orientations were kept separate.

But if you want, I could still remove this part and wait that things get more confusion-free regarding orientations.

DominicDirkx commented:
Since I'm not quite sure why/how this works the way it does, I think it's best that we remove it for now. This type of mess is also a clear reason why we should refactor the whole orientation part :)

gaffarelj commented:
Perfect, I commented this part out, and just merged this pull request.
The thrust guide is now available on tudat-space [here](https://tudat-space.readthedocs.io/en/latest/_src_user_guide/state_propagation/propagation_setup/acceleration_models/thrust.html).

@FilippoOggionni as we discussed on Slack, feel free to send me comments on this guide if you wish.

FilippoOggionni commented on docs/source/_src_user_guide/state_propagation/propagation_setup/acceleration_models/thrust.rst:
```diff
@@ -0,0 +1,301 @@
+Thrust direction from existing orientation
+==========================================
+
+The orientation of the vehicle is in some cases already defined. This could be thanks to aerodynamic guidance or to the propagation of rotational dynamics. In this context, the thrust direction can be computed from the body-fixed direction. 
```
> In this context, the thrust direction can be computed from the body-fixed direction.

Do you mean "body-fixed orientation"? I am not an expert about thrust in tudat and from reading through this page I get a bit confused between orientation and direction. Let's try to be careful in using the correct terms (perhaps this is already the case, not sure) to avoid confusion for the users

gaffarelj replied:
I do, indeed.
I think it's normal that you are confused, to be honest, the current implementation is confusing I would say (though it does work).
Referring to my previous comment, we do seek to make this less confusing in the future.

gaffarelj replied:
And I totally agree that, being myself still partially confused by this orientation/direction definition (especially combined with aerodynamic orientation or rotational dynamics), I should make even more of an effort to not transmit this confusion to users 🙂

FilippoOggionni commented on docs/source/_src_user_guide/state_propagation/propagation_setup/acceleration_models/thrust.rst:
```diff
@@ -0,0 +1,301 @@
+The direction of the thrust in the body-fixed frame can be additionally defined when specifying the `thrust magnitude <#thrust-magnitude>`_ (note that this design is currently under review, and may well be refactored in the near future).
+
+
+Thrust direction from state guidance settings
```
If the code snippets are removed, I'd consider changing all the ways to create the thrust direction (and also magnitude) simply to bold font instead of actual subsections. I'd do that, adding a link to the related entry in the API docs and a few lines of explanation, not more than that, to keep everything a bit more compact.

FilippoOggionni commented on docs/source/_src_user_guide/state_propagation/propagation_setup/acceleration_models/thrust.rst:
```diff
@@ -0,0 +1,301 @@
+~~~~~~~~~~~~~~~~
+
+Below, the different methods that have been implemented to define the direction of the thrust are outlined. Note that these settings are *only* relevant if you use the :func:`~tudatpy.numerical_simulation.propagation_setup.acceleration.thrust_from_direction_and_magnitude` function.
+
```
I'd specify here that all these methods return a `ThrustDirectionSettings` object.

FilippoOggionni commented on docs/source/_src_user_guide/state_propagation/propagation_setup/acceleration_models/thrust.rst:
```diff
@@ -0,0 +1,301 @@
+~~~~~~~~~~~~~~~~
+
+Below, the different methods that have been implemented to define the magniude of the thrust are outlined. Note that these settings are *only* relevant if you use the :func:`~tudatpy.numerical_simulation.propagation_setup.acceleration.thrust_from_direction_and_magnitude` function.
+	
```
I'd specify here that all these methods return a `ThrustMagnitudeSettings` object.

FilippoOggionni reviewed (commented):
Great job overall!

gaffarelj commented on docs/source/_src_user_guide/state_propagation/propagation_setup/acceleration_models/thrust.rst:
```diff
@@ -0,0 +1,301 @@
+
+Below, the different methods that have been implemented to define the direction of the thrust are outlined. Note that these settings are *only* relevant if you use the :func:`~tudatpy.numerical_simulation.propagation_setup.acceleration.thrust_from_direction_and_magnitude` function.
+
+In all of these methods, the thrust direction that is defined is always in the **inertial frame**, either directly or indirectly. It is important to realize that, when specifying a thrust direction, the vehicle orientation itself is automatically defined.
```
(I don't think that I will do anything regarding this comment.)
What this means is that, when the thrust direction is defined in Tudat, the body direction is automatically defined at the same time, based on this thrust direction.
But there is still a big general confusion about defining orientation, and this will potentially be changed in the (near?) future, rendering part of this guide obsolete anyways.

gaffarelj commented:
Thanks a lot for the comments and commits you made to this guide @FilippoOggionni!
I have overall incorporated all of your comments from above, I have removed all of the smaller code snippets (I kept the more important ones), and I have added these to the API docs (still to be built later on).

FilippoOggionni commented:
No worries, it looks great! Awesome job 🚀

DominicDirkx commented:
Awesome, thanks!
